experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -89,112 +94,114 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
-                                       Operation *currOp) {
-  // End traversal if currOp is already in visited set
-  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
-    return false;
-
-  if (isa<handshake::SpecCommitOp>(currOp)) {
-    // End traversal at Commits and notify that the path leads to a commit
-    markedPath.insert(currOp);
-    return true;
-  } else {
-    bool foundCommit = false;
-    // Continue DFS traversal
-    for (Operation *succOp : currOp->getUsers())
-      foundCommit |= markPathToCommitsRecursive(visited, markedPath, succOp);
-
-    if (foundCommit)
-      markedPath.insert(currOp);
-
-    return foundCommit;
-  }
-}
-
-// Mark all paths that lead to commit units by adding them to markedPath
-static void markPathToCommits(llvm::DenseSet<Operation *> &markedPath,
-                              SpeculatorOp &specOp) {
-  // Create visited set
-  llvm::DenseSet<Operation *> visited;
-  visited.insert(specOp);
-
-  // Traverse IR starting at the speculator's output
-  for (Operation *succOp : specOp.getDataOut().getUsers())
-    markPathToCommitsRecursive(visited, markedPath, succOp);
-}
+// The list item to trace the branches that need to be replicated
+struct BranchTracingItem {
+  // A value whose spec tag is used to discard non spec case
+  Value valueForSpecTag;
+  // The branch to replicate
+  handshake::ConditionalBranchOp branchOp;
+  // The index of the result of the branchOp
+  unsigned branchDirection;
+};
 
-// This recursive function traverses the IR along a marked path and creates
-// a control path by replicating the branches it finds in the way. It stops
-// at commits and connects them to the newly created path with value
-// ctrlSignal
-void HandshakeSpeculationPass::routeCommitControl(
-    llvm::DenseSet<Operation *> &markedPath, Value ctrlSignal,
-    OpOperand &currOpOperand) {
+// This function traverses the IR and creates a control path by replicating the
+// branches it finds in the way. It stops at commits and connects them to the
+// newly created path with value ctrlSignal
+static void routeCommitControlRecursive(
+    MLIRContext *ctx, SpeculatorOp &specOp,
+    llvm::DenseSet<Operation *> &arrived, OpOperand &currOpOperand,
+    const std::list<BranchTracingItem> &branchTracingList) {
   Operation *currOp = currOpOperand.getOwner();
-  // End traversal if currOp is not in the marked path to commits
-  if (!markedPath.contains(currOp))
+
+  // End traversal if currOpOperand is already arrived
+  if (arrived.contains(currOp))
     return;
+  arrived.insert(currOp);
 
-  // Remove operation from the set to avoid visiting it twice
-  markedPath.erase(currOp);
+  // If the traversal reaches a speculator or a save commit unit, stop it
+  // All commit units should be found within the same loop iteration
+  // And passing through these units may result in the replication of redundant
+  // branches
+  if (isa<handshake::SpeculatorOp>(currOp))
```

Comment:
I think change "stop it" to "end traversal"
I think replace 
"//All commit units should be found within the same loop iteration
  // And passing through these units may result in the replication of redundant
  // branches"
with
"In order to route the commit control, we search the speculative region from top to bottom.
The speculative region begins at the speculator output, and ends when either the speculator, a save-commit unit, or a commit unit is encountered.
Any path from the output of the speculator to a commit unit needs a corresponding control path.
Paths that do not reach commit units are handled elsewhere"

It's a bit wordy, but also I think fully explains it